proposals/0000-unsaturated-type-families.rst

```diff
@@ -0,0 +1,656 @@
+enabled with the pragma ``{-# LANGUAGE UnsaturatedTypeFamilies #-}``.
+The pragma implies ``TypeFamilies``.
+
+.. _Syntax:
+
+Syntax changes
+~~~~~~~~~~~~~~
+
+GHC's parser includes the following production rules for types::
+
+  type ::= btype '->' ctype
+       | ...
+
+
+  ctype ::= 'forall' tv_bndrs '->' ctype
+        |   'forall' tv_bndrs '.' ctype
+        |   'context '=>' ctype
+
+This proposal adds the following rules::
+
+  type ::= btype '->' ctype
+       |   btype '->' PREFIX_AT atype ctype
+       | ...
+
+
+  ctype ::= 'forall' tv_bndrs '->' ctype
+        |   'forall' tv_bndrs '->' PREFIX_AT atype ctype
+        |   'forall' tv_bndrs '.' ctype
+        |   'forall' tv_bndrs '.' PREFIX_AT atype ctype
+        |   'context '=>' ctype
+        |   'context '=>' PREFIX_AT atype ctype
+        | ...
+
+Where ``PREFIX_AT`` stands for the lexer token ``@`` that is to be parsed as a
+prefix operator.
+
+That is, it is now possible to annotate each existing form of quantifier with
+matchability information.
+
+Matchabilities
+~~~~~~~~~~~~~~
+
+Matchability is a first-class type, and is defined in ``GHC.Exts`` as ::
+
+  data Matchability = Matchable | Unmatchable
+
+We use these long names to improve the discoverability of the feature,
+but also provide shorter synonyms, which are used in the example above::
+
+  type M = 'Matchable
+  type U = 'Unmatchable
+
+The ``Matchability`` type and the ``M`` and ``U`` synonyms are exported from
+the ``GHC.Exts`` module.
+
+FUN
+~~~
+
+The full kind of the ``(->)`` constructor becomes ::
+
+  (->) :: forall (m :: Matchability)
+                 {q :: RuntimeRep} {r :: RuntimeRep}.
+          TYPE q -> TYPE r -> Type
+
+The matchability part of the arrow can be instantiated using visible type
+application in types, a recent addition to GHC.
+
+The ``a -> @m b`` syntax is thus syntactic sugar for ``(->) @m a b``.
+
+Since the ``LinearTypes`` extension has landed in GHC, the `(->)` constructor is defined
+as a synonym for a more general constructor ``FUN`` that takes a multiplicity
+argument. The full kind of ``FUN`` under the current proposal becomes ::
+
+  type FUN :: forall (m :: Matchability).
+              forall (n :: Multiplicity) ->
+              forall {q :: RuntimeRep} {r :: RuntimeRep}. TYPE q -> TYPE r -> Type
+
+which now accounts for both matchability and multiplicity annotations.
+Then ``(->)`` is defined morally as::
+
+  type (->) @m = FUN @m 'Many
+
+Since the matchability argument is invisible, this synonym works just like one
+would expect (in particular, there's no unexpected interaction from the fact
+that ``(->)`` needs to bind the matchability argument to apply it out-of-order.
+
+Note that the matchability argument is invisible, therefore manually specifying
+it is optional.
+
+Inference
+~~~~~~~~~
+
+The meaning of unannotated ``forall``s and ``->`` is inferred, using the
+following rules (for more details see the *Overview* section):
```

Comment:
It will - when no arrow was written!

```
type family Map f xs where
  Map f '[] = '[]
  Map f (x ': xs) = f x ': Map f xs
```

will be inferred (when polykinds is on)

```
Map :: (a -> @m b) -> @U [a] -> @U [b]
```

This simply follows the no-cusk generalisation strategy and nothing clever happens here.

I'll update the text to reflect this.